File: Data Structures & Algorithms/count-number-of-islands/submission-0.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0

        rows, cols = len(grid), len(grid[0])
        visited = set()
        res = 0

        def bfs(r, c):
            queue = deque()
            queue.append((r, c))
            visited.add((r, c))
            directions = [(0, 1),(0, -1),(1, 0),(-1, 0)]

            while queue:
                row, col = queue.popleft()
                for direction in directions:
                    cur_r, cur_c = row + direction[0], col + direction[1]
                    if (
                        0 <= cur_r < rows and
                        0 <= cur_c < cols and
                        grid[cur_r][cur_c] == "1" and
                        (cur_r, cur_c) not in visited
                    ):
                        queue.append((cur_r, cur_c))
                        visited.add((cur_r, cur_c))

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1" and (r, c) not in visited:
                    bfs(r, c)
                    res += 1
        
        return res

File: Data Structures & Algorithms/valid-sudoku/submission-0.cpp
class Solution {
    public:
        bool isValidSudoku(vector<vector<char>>& board) {
            if (board.empty()){
                return false;
            }
            vector<vector<bool>> row (9, vector<bool>(9, false));
            vector<vector<bool>> col (9, vector<bool>(9, false));
            vector<vector<bool>> grid(9, vector<bool>(9, false));
            for (int r = 0; r < board.size(); r++){
                for (int c = 0; c < board[0].size(); c++){
                    char cur_ch = board[r][c];
                    if (cur_ch == '.'){
                        continue;
                    }
                    if ((row[r][cur_ch - '1']) || (col[c][cur_ch - '1']) || (grid[((r / 3) * 3) + (c / 3)][cur_ch - '1'])){
                        return false;
                    }
                    row[r][cur_ch - '1'] = true;
                    col[c][cur_ch - '1'] = true;
                    grid[((r / 3) * 3) + (c / 3)][cur_ch - '1'] = true;
                }
            }
            return true;
        }
    };